### PR TITLE
WORKSHOP clean up optional dependency imports

### LINKED ISSUE #338: Clean up optional dependencies (mostly in readers I think)
# Requested Update

### Description
There are a number of "optional" dependencies that look like this:

```python
try:
    from pyhdf.HDF import ishdf
except ImportError:
    LOG.info(
        "Failed import pyhdf in /readers/modis_hdf4.py. "
        + "If you need it, install it."
    )
```

Since these occur relatively frequently, I'd like to manage them a little more cleanly. To do this, we can use a context manager like this:

```python
import logging
from contextlib import contextmanager
 
LOG = logging.getLogger(__name__)
 
 
@contextmanager
def import_optional_dependences(importer):
    """Attempt to import a package and print to LOG.info if the import fails."""
    try:
        yield None
    except ImportError as err:
        LOG.info(f'Failed to import {err.name} in {importer}. If you need it, install it.')
```

This gets used like this:

```python
from testing.context_manager import import_optional_dependences
from geoips.commandline.log_setup import setup_logging
 
LOG = setup_logging(logging_level="INFO")                                                                                                                                                                                                                                                                                   
 
 
with import_optional_dependences(__file__):
    import blahblahblah
    import numexpr
```

### Background and Motivation
*  [ OPTIONAL: Provide additional detailed information and background related to this issue ]

### Alternative Solutions
*  [ OPTIONAL: If appropriate, list alternative fixes that could be considered]
  
### Environment
*  [ IF APPROPRIATE: Corresponding repository versions ]

### Code to demonstrate issue
*  [ IF APPROPRIATE: Can include steps here to identify/demonstrate bug fix or desired feature ]

### Checklist for Completion
* [ ] <!--- Include a list of steps that must be complete in order to close this Issue --->

### PR DESCRIPTION
**Pull request by evrose54**
# Reviewer Checklist

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] NO REQUIRED ***unit tests*** (explain why not required)
* [x] NO REQUIRED ***integration tests*** (explain why not required)
* [x] Required ***documentation*** added for new/modified functionality
* [x] Required ***release notes*** added for new/modified functionality
* [ ] Required ***updates to other repos*** complete

# Related Issues
fixes NRLMMD-GEOIPS/geoips#338
See [this PR](https://github.com/NRLMMD-GEOIPS/geoips/pull/438) for the GeoIPS implementation of this PR. Updating other repos as necessary to reflect this new implementation. Make sure to merge that branch first before merging this one.

# Testing Instructions
Run geoips/tests/integration_tests/full_test.sh to ensure these changes have passed. They passed on my end.

# Summary
*From GEOIPS#338: 2023-07-19, Clean up optional dependencies*

Some GeoIPS readers include optional dependency statements that are required to read
certain file formats. While keeping these imports as optional is ok, we should clean up
the manner in which this is implemented. To do so, we've created a separate
geoips.testing.context_manager.py script which can handle optional imports scattered
throughout the GeoIPS codebase. This is essentially replacing our old manner of optional
dependencies with a new method that keeps things clean.

    - modified: geoips/plugins/modules/procflows/data_fusion.py

### CODE REVIEW

File: docs/source/releases/v1_12_0a0.rst [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ | # # # the terms of the NRLMMD License included with this program. This program is
+ | # # # distributed WITHOUT ANY WARRANTY; without even the implied warranty of
+ | # # # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the included license
+ | # # # for more details. If you did not receive the license, for more information see:
+ | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
+
+Version 1.11.7a0 (2023-10-06)
+*****************************
+
+* Refactor: Clean up optional dependencies throughout code
```
Comment (evrose54) [inline docs/source/releases/v1_12_0a0.rst]: Don't merge this until [this PR](https://github.com/NRLMMD-GEOIPS/geoips/pull/438) has been merged.

**Review by jsolbrig — CHANGES_REQUESTED** [review]
@evrose54 I'm running into a problem with this one. When I try to run `create_plugin_registires` I get the following:

```
> create_plugin_registries 
15_183838 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'interp_scipy' from 'geoips', interface_name is 'None'
15_183838 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'boxdefinitions' from 'geoips', interface_name is 'None'
15_183838 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'interp_pyresample' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'geostationary_geolocation' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'hrit_reader' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'remss_reader' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:792  INTERACTIVE: Skipping module 'tc_file_naming' from 'geoips', interface_name is 'None'
15_183839 create_plugin_registries.py:296  INTERACTIVE: Writing /local/home/jsolbrig/geoips/github/geoips/geoips/registered_plugins.json
15_183839 create_plugin_registries.py:296  INTERACTIVE: Writing /local/home/jsolbrig/geoips/github/geoips_clavrx/geoips_clavrx/registered_plugins.json
Traceback (most recent call last):
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/utils/context_managers.py", line 21, in import_optional_dependencies
    yield None
  File "/local/home/jsolbrig/geoips/github/data_fusion/data_fusion/plugins/modules/procflows/data_fusion.py", line 50, in <module>
    from geoips_db.utils.database_writes import write_stats_to_database
ModuleNotFoundError: No module named 'geoips_db'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jsolbrig/anaconda3/envs/geoips-github/bin/create_plugin_registries", line 8, in <module>
    sys.exit(main())
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/create_plugin_registries.py", line 881, in main
    create_plugin_registries(plugin_packages, save_type)
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/create_plugin_registries.py", line 388, in create_plugin_registries
    error_message += parse_plugin_paths(plugin_paths, package, pkg_dir, plugins)
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/create_plugin_registries.py", line 471, in parse_plugin_paths
    error_message += add_module_plugin(
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/create_plugin_registries.py", line 748, in add_module_plugin
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/local/home/jsolbrig/geoips/github/data_fusion/data_fusion/plugins/modules/procflows/data_fusion.py", line 48, in <module>
    with import_optional_dependencies(__file__):
  File "/home/jsolbrig/anaconda3/envs/geoips-github/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/local/home/jsolbrig/geoips/github/geoips/geoips/utils/context_managers.py", line 28, in import_optional_dependencies
    getattr(LOG, loglevel)(err_str)
AttributeError: 'Logger' object has no attribute '/local/home/jsolbrig/geoips/github/data_fusion/data_fusion/plugins/modules/procflows/data_fusion.py'
```

This looks like it is likely a problem in the `import_optional_dependencies` context manager.

**Comment by evrose54** [comment]
Weird @jsolbrig. I'm not running into this problem at all. Could run ``create_plugin_registries`` on this PR again just to make sure the problem is still occurring on your end?

**Comment by jsolbrig** [comment]
Weirdly, I no longer see this issue. This appears to be ready. I'm going to force the CI to re-run, then mark this as ready to merge.

File: docs/source/releases/v1_12_0a0.rst (hunk 1 above)
Comment (jsolbrig) [reply docs/source/releases/v1_12_0a0.rst]: The parent PR has been merged.

**Review by jsolbrig — CHANGES_REQUESTED** [review]
This is ready to go, but the release notes need to be updated to the new format when we're able to do so.

**Comment by evrose54** [comment]
Great; Thanks for trying this out again! I wonder what the problem was, but I'm glad to see this is no longer an issue.

**Review by jsolbrig — DISMISSED** [review]

File: docs/source/releases/HEADER [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+ | # # #
+ | # # # Author:
+ | # # # Naval Research Laboratory, Marine Meteorology Division
+ | # # #
+ | # # # This program is free software: you can redistribute it and/or modify it under
+ | # # # the terms of the NRLMMD License included with this program. This program is
+ | # # # distributed WITHOUT ANY WARRANTY; without even the implied warranty of
+ | # # # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the included license
+ | # # # for more details. If you did not receive the license, for more information see:
+ | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
```
Comment (mindyls) [inline docs/source/releases/HEADER]: Is this for generating the automated release notes?  Should we use the updated header?

Comment (jsolbrig) [reply docs/source/releases/HEADER]: Yes and yes! This should use the new header.

Comment (mindyls) [reply docs/source/releases/HEADER]: You don't want the long one?

Comment (mindyls) [reply docs/source/releases/HEADER]: Done

File: docs/source/releases/HEADER [hunk 3]
```diff
@@ -0,0 +1,4 @@
+.. dropdown:: Distribution Statement
+
+ | # # # This source code is protected under the license referenced at
+ | # # # https://github.com/NRLMMD-GEOIPS.
```
Comment (mindyls) [inline docs/source/releases/HEADER]: Is this still required?  Is this part of brassy?

File: docs/source/releases/v1_12_0a0.rst [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ | # # # the terms of the NRLMMD License included with this program. This program is
+ | # # # distributed WITHOUT ANY WARRANTY; without even the implied warranty of
+ | # # # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the included license
+ | # # # for more details. If you did not receive the license, for more information see:
+ | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
+
+Version 1.12.0a0 (2024-02-12)
+*****************************
+
+No updates.
```
Comment (mindyls) [inline docs/source/releases/v1_12_0a0.rst]: Is there a reason 1.12.0a0 and 1.13.0a0 are both added here with "No updates" ?

File: docs/source/releases/v1_13_0a0.rst [hunk 5]
```diff
@@ -12,48 +12,7 @@
  | # # # for more details. If you did not receive the license, for more information see:
  | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
 
-Version 1.13.0a0 (2024-06-07)
+Version 1.13.0a0 (2024-07-10)
```
Comment (mindyls) [inline docs/source/releases/v1_13_0a0.rst]: What happened to these updates?

File: docs/source/releases/v1_13_0a0.rst [hunk 6]
```diff
@@ -1,16 +1,7 @@
...
- | # # # Naval Research Laboratory, Marine Meteorology Division
- | # # #
- | # # # This program is free software: you can redistribute it and/or modify it under
- | # # # the terms of the NRLMMD License included with this program. This program is
- | # # # distributed WITHOUT ANY WARRANTY; without even the implied warranty of
- | # # # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the included license
- | # # # for more details. If you did not receive the license, for more information see:
- | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
+ | # # # This source code is protected under the license referenced at
+ | # # # https://github.com/NRLMMD-GEOIPS.
```
Comment (mindyls) [inline docs/source/releases/v1_13_0a0.rst:4]: NOTE: For some reason all of the contents of this release note had been removed.  I put the contents back - I don't know why the contents had been removed, but they seem to still be relevant (matplotlib_linear_norm -> linear_norm, and updating test scripts to geoips run data_fusion. Note I cahnged linear_norm BACK to matplotlib_linear_norm in 1.13.0)

File: docs/source/releases/latest/666-final-bug-fixes-prior-to-workshop.yaml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+documentation:
+- description: |
+    Add create_plugin_registries and missing test data to README
+  title: 'Add create_plugin_registries and missing test data to README'
+  related-issue:
+    number: 666
+    url: 'https://github.com/NRLMMD-GEOIPS/geoips'
+  files:
+    modified: ['README.md']
```
Comment (mindyls) [inline docs/source/releases/latest/666-final-bug-fixes-prior-to-workshop.yaml]: This was from a previous PR, before we updated the YAML release note formatting.

File: docs/source/releases/index.rst [hunk 8]
```diff
@@ -43,6 +34,7 @@ Version 1.12
 
    v1_12_4
    v1_12_3
+   v1_12_0a0
```
Comment (mindyls) [inline docs/source/releases/index.rst]: Why did we add a release note for a release that appears to have no updates?

File: docs/source/releases/HEADER (hunk 3 above)
Comment (mindyls) [reply docs/source/releases/HEADER]: Does it do anything?

File: docs/source/releases/v1_12_0a0.rst (hunk 4 above)
Comment (mindyls) [reply docs/source/releases/v1_12_0a0.rst]: I put 1.13.0a0 back to its original contents. Can we just delete 1.12.0a0?

File: docs/source/releases/HEADER (hunk 3 above)
Comment (evrose54) [reply docs/source/releases/HEADER]: I think we can get rid of that. I added that when initially testing brassy as I thought we needed it but after talking to Gwyn it sounds like we don't.

File: docs/source/releases/v1_14_0a0.rst
Comment (mindyls) [inline docs/source/releases/v1_14_0a0.rst:1]: Note this was removed, and the only remaining update moved to latest/666...yaml.  Now we can auto-create new release note.

File: docs/source/releases/v1_12_0a0.rst (hunk 4 above)
Comment (evrose54) [reply docs/source/releases/v1_12_0a0.rst]: Yup! I think this branch was made before any updates happened in v1_13_0a0. Not sure about the no updates part. I thought that was from you but maybe it was from me. Can't remember. I'll remove it nonetheless.

File: docs/source/releases/index.rst (hunk 8 above)
Comment (evrose54) [reply docs/source/releases/index.rst]: I'll remove that.

File: docs/source/releases/v1_13_0a0.rst (hunk 6 above)
Comment (evrose54) [reply docs/source/releases/v1_13_0a0.rst:4]: Sounds good! Thanks for doing that. I'm not sure how the release notes were removed either. This branch was created a while ago and that might have something to do with it.

**Review by jsolbrig — APPROVED** [review]

File: docs/source/releases/index.rst [hunk 9]
```diff
@@ -1,31 +1,14 @@
...
 *************
 
-Version 1.14
-------------
-
-.. toctree::
-   :maxdepth: 1
-
-   v1_14_0a0
-
```
Comment (mindyls) [inline docs/source/releases/index.rst:28]: Do not merge until docs/source/releases/*.rst check is failing.

**Review by mindyls — APPROVED** [review]
Once the docs/source/releases/*.rst check is correctly fialing, I'm good to merge this.

File: docs/source/releases/index.rst (hunk 9 above)
Comment (mindyls) [reply docs/source/releases/index.rst:28]: image

**Comment by mindyls** [comment]
image

Overriding and merging - intentionally updated docs/soure/releases/*.rst files

**Pull request merged**